### PR TITLE
Replace valgrind by all sanitizers available.

### PR DESCRIPTION
**Pull request by tcojean**
This tentatively removes valgrind and instead add leak sanitizer and memory sanitizer check. The hope is to gain quite a CI speedup without losing in quality checks (in fact, UBsan could find new important issues).

+ Add a new UBSAN configuration for undefined behavior.
+ Add a new leak sanitizer configuration.
+ Add a sanitizers test to ensure all sanitizers work.
+ Add a new ginkgo_create_thread_test helper for the sanitizers test.
+ Add CI steps to check for leaks and undefined behavior.

Also update containers with various fixes and additions:

+ Add Metis build option
+ Add MPI build option
+ Make CUDA 10.1 to be based on ubuntu 18.04.
+ Fix an issue with HIP and CUDA <= 9.1

### CODE REVIEW

File: cmake/build_type_helpers.cmake [hunk 1]
```diff
@@ -49,8 +55,14 @@ GKO_XCOMPILER(${PROJECT_NAME}_NVCC_COVERAGE_COMPILER_FLAGS "-g;-O0;--coverage")
 GKO_XCOMPILER(${PROJECT_NAME}_NVCC_COVERAGE_LINKER_FLAGS   "--coverage")
 GKO_XCOMPILER(${PROJECT_NAME}_NVCC_TSAN_COMPILER_FLAGS     "-g;-O1;-fsanitize=thread;-fno-omit-frame-pointer;-fPIC")
 GKO_XCOMPILER(${PROJECT_NAME}_NVCC_TSAN_LINKER_FLAGS       "-fsanitize=thread;-static-libtsan -fno-omit-frame-pointer -fPIC")
```
[the remark this answers is not in the document]
Comment (tcojean) [reply cmake/build_type_helpers.cmake]: You're right. I'll check these.

**Comment by tcojean** [comment]
@pratikvn The sanitizers do not work with CUDA or HIP code. In addition, I tried to make CTest interact with cuda-memcheck to create a custom step which would take reports from that tool, but was never successful in doing so.

**Comment by tcojean** [comment]
I needed to update the CUDA containers for this PR. Looks like that created an issue with CUDA + Intel for everyone. I will try to fix this.

**Comment by tcojean** [comment]
Hopefully everything is now finally working. Still awaiting normal CI confirmation. I had to change the containers settings because UBSAN works only with Ubuntu 18.04 but using Ubuntu 18.04 is extremely complex as it conflicts with some CUDA, Intel, etc versions.

The new sanitizer setup is in place and working. You can see what it looks like here:
https://my.cdash.org/index.php?project=Ginkgo+Project

I added four new tests which are designed to create an error for every of the new sanitizers. This is a way to test the sanitizers, so that we know they are properly setup and actually working. This proved to be important as UBSAN was clearly not working for most of the time and debugging this issue proved to be *very* time consuming. Here are the new added tests:
+ A wrong address access,
+ A memory leak,
+ A threading race condition,
+ An undefined integer operation.

**Comment by tcojean** [comment]
There is still a problem with one setup, clang+CUDA+HIP compilation for some reason. Please do not merge anything before I fix it:
https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/472997356

**Comment by tcojean** [comment]
Everything is now fixed. For some reason, gtest is not compiled with `-fPIC` and that never was an issue before but it is for HIP + CUDA 10.1 + Ubuntu 18.04...

File: dev_tools/containers/ginkgo-cuda-base.py [hunk 2]
```diff
@@ -114,8 +123,13 @@
...
                           'cmake ..', 'make install'])
 Stage0 += shell(commands=['rm -rf /var/tmp/HIP'])
-Stage0 += shell(commands=['cd /var/tmp',
-                          'git clone https://github.com/tcojean/hipBLAS.git'])
+if float(cuda_version) >= float(9.2):
+    Stage0 += shell(commands=['cd /var/tmp',
+                              'git clone https://github.com/tcojean/hipBLAS.git'])
+else:
+    Stage0 += shell(commands=['cd /var/tmp',
+                              'git clone https://github.com/tcojean/hipBLAS.git --branch master-rocm-2.10'])
```
[the remark this answers is not in the document]
Comment (tcojean) [reply dev_tools/containers/ginkgo-cuda-base.py]: Yes it is, AFAIK, newer rocm versions do not support older CUDA versions, so we have to stick with this version of hipblas (2.10) for cuda < 9.2. I need to update my forks again though, maybe the situation changed in between.

File: cmake/CTestScript.cmake [hunk 3]
```diff
@@ -51,7 +54,8 @@
 #
 # ``CTEST_BUILD_CONFIGURATION``
 # Which configuration should Ginkgo be built with. Default `DEBUG`.
-# The supported values are: COVERAGE, ASAN, TSAN, DEBUG and RELEASE.
+# The supported values are: COVERAGE, TSAN, UBSAN DEBUG and
```
[the remark this answers is not in the document]
Comment (tcojean) [reply cmake/CTestScript.cmake]: Thanks!

File: dev_tools/containers/ginkgo-cuda-base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yhmtsai) [reply dev_tools/containers/ginkgo-cuda-base.py]: Got it. Thanks

**Pull request merged**